client/mm/exchange_adaptor.go

```diff
@@ -2638,6 +2498,412 @@ func (u *unifiedExchangeAdaptor) handleDEXNotification(n core.Notification) {
+
+	additionalBaseFees, additionalQuoteFees := perLot.baseFunding, perLot.quoteFunding
+	if u.baseID == u.quoteFeeID {
+		additionalBaseFees += perLot.baseRedeem * dexBuyLots
+	}
+	if u.quoteID == u.baseFeeID {
+		additionalQuoteFees += perLot.quoteRedeem * dexSellLots
+	}
+	var baseAvail, quoteAvail uint64
+	if baseInv.total > additionalBaseFees {
+		baseAvail = baseInv.total - additionalBaseFees
+	}
+	if quoteInv.total > additionalQuoteFees {
+		quoteAvail = quoteInv.total - additionalQuoteFees
+	}
+
+	// matchability is the number of lots that can be matched with a specified
+	// asset distribution.
+	matchability := func(dexBaseLots, dexQuoteLots, cexBaseLots, cexQuoteLots uint64) uint64 {
+		sells := utils.Min(dexBaseLots, cexQuoteLots, maxSellLots)
+		buys := utils.Min(dexQuoteLots, cexBaseLots, maxBuyLots)
+		return buys + sells
+	}
+
+	currentScore := matchability(baseInv.dexLots, quoteInv.dexLots, baseInv.cexLots, quoteInv.cexLots)
+
+	// targetedSplit finds a distribution that targets a specified ratio of
+	// dex-to-cex.
+	targetedSplit := func(avail, dexTarget, cexTarget, dexLot, cexLot uint64) (dexLots, cexLots, extra uint64) {
+		if dexTarget+cexTarget == 0 {
+			return
+		}
+		cexR := float64(cexTarget*cexLot) / float64(dexTarget*dexLot+cexTarget*cexLot)
+		cexLots = uint64(math.Round(cexR*float64(avail))) / cexLot
+		cexBal := cexLots * cexLot
+		dexLots = (avail - cexBal) / dexLot
+		dexBal := dexLots * dexLot
+		if cexLot < dexLot {
+			cexLots = (avail - dexBal) / cexLot
+			cexBal = cexLots * cexLot
+		}
+		extra = avail - cexBal - dexBal
+		return
+	}
+
+	baseSplit := func(dexTarget, cexTarget uint64) (dexLots, cexLots, extra uint64) {
+		return targetedSplit(baseAvail, utils.Min(dexTarget, maxSellLots), utils.Min(cexTarget, maxBuyLots), perLot.dexBase, perLot.cexBase)
+	}
+	quoteSplit := func(dexTarget, cexTarget uint64) (dexLots, cexLots, extra uint64) {
+		return targetedSplit(quoteAvail, utils.Min(dexTarget, maxBuyLots), utils.Min(cexTarget, maxSellLots), perLot.dexQuote, perLot.cexQuote)
+	}
+
+	// We'll keep track of any distributions that have a matchability score
+	// better than the score for the current distribution.
+	type scoredSplit struct {
+		score uint64 // matchability
+		// spread is just the minimum of baseDeposit, baseWithdraw, quoteDeposit
+		// and quoteWithdraw. This is tertiary criteria for prioritizing splits,
+		// with a higher spread being preferable.
+		spread                      uint64
+		fees                        uint64
+		baseDeposit, baseWithdraw   uint64
+		quoteDeposit, quoteWithdraw uint64
+	}
+	baseSplits := [][2]uint64{
+		{baseInv.dex, baseInv.cex},   // current
+		{dexSellLots, dexBuyLots},    // ideal
+		{quoteInv.cex, quoteInv.dex}, // match the counter asset
+	}
+	quoteSplits := [][2]uint64{
+		{quoteInv.dex, quoteInv.cex},
+		{dexBuyLots, dexSellLots},
+		{baseInv.cex, baseInv.dex},
+	}
```

Comment:
What's the reasoning behind these splits? Couldn't we try all combinations from 0 lots on dex and all on cex to all on dex and 0 on cex. If the total balance is > 100 lots, then we could do 1% of balance on dex and 99% on cex, and modify by 1% on each iteration, to avoid the check taking too long. Seems like this 3x3 set would miss many superior combinations.

Reply:
For some markets, orders might be 100,000 lots or more. And every one of those would be compared with with each of 100,000 or so on the other side, giving 1,000,000,000,000 possible combos. We could potentially do some kind of nested binary search, but the information we're looking for doesn't lend itself well to bisecting. 
These targets are reasonable and give snappy results, and the 20 combinations targeted satisfy all of my test conditions.

Reply:
If it's over 100 lots, for example 100,000 lots like you said, we could divide the total by 100 and check in increments of 1%, 1000 lots in this case. Then the most possible combinations would be 100^2 = 10,000, which is probably still snappy. Maybe it's worthless though. I'll play around with some scenarios.

Reply:
Instead of having "extra" we could try each combination in increments of 1% from 0 to the total balance of an asset across the cex and dex. For scoring, the first criteria would be to be able to place as many of the necessary lots as possible, but if we are able to place them all, the second criteria would be to maximize the amount of lots that can be traded in either direction before we once again are unable to place the necessary lots. This would minimize the chances of a future rebalance being required.

Reply:
I'm definitely not opposed to any implementation that provides a net benefit. I'm not great with big O notation, but I think mine is O(1), nested binary is O((log n)^2), and the one you're proposing is O(n^2). Like you said though, If we can ensure somehow that our upper bound is still reasonably fast, I don't really care how its done.